src/VecSim/vec_sim_index.h

```diff
@@ -55,4 +57,37 @@ struct VecSimIndexAbstract : public VecSimIndexInterface {
     // Static class functions
     static timeoutCallbackFunction timeoutCallback;
     static void setTimeoutCallbackFunction(timeoutCallbackFunction callback);
+
+    template <typename DataType>
+    static void NormalizeVector(const void *input_vector, size_t dim, DataType *output = NULL);
 };
+
+static const double MAX_FP64 = std::numeric_limits<double>::max();
+
+template <typename DataType>
+void VecSimIndexAbstract::NormalizeVector(const void *input_vector, size_t dim, DataType *output) {
+
+    // if output is not NULL and doesn't point to the same memory area as input.
+    if (output && output != (DataType *)input_vector) {
+        memcpy(output, input_vector, dim * sizeof(DataType));
+    } else { // else we are normalizing inplace.
+        output = (DataType *)input_vector;
+    }
+    double sum = 0;
+    for (size_t i = 0; i < dim; i++) {
+        double to_add = (double)output[i] * (double)output[i];
+        // Protect from overflow.
+        assert(MAX_FP64 - sum >= to_add);
+        sum += to_add;
```

Comment:
Why the double add? Just check for overflow.
Check those links https://stackoverflow.com/questions/15655070/how-to-detect-double-precision-floating-point-overflow-and-underflow
https://en.cppreference.com/w/cpp/numeric/fenv/fetestexcept

Reply:
I tried a few kinds of library functions that check overflow and I've found that they might not be good enough,
due to the floating point rounding mechanism which depends on the value:
https://stackoverflow.com/questions/37830178/why-no-overflow-when-using-double-variable
So assuming we have reached the maximum value of fp32, and we keep adding reasonable values, these flags won't be set...

The solution here is to check overflow before adding